SunbirdAI/noise-sensors-monitoring
# Device configuration

### What does this PR do?
- Adds the device configuration fields to the Device model and displays the Device configuration on the DeviceDetail page
- Adds a form to edit the device configuration
- Retrieves the Device Configuration for use in the MQTT app
- Sets up the subscription and publication code necessary for communicating with the broker to send the Device Configuration.
- Sets up a separate container for the MQTT app.
- Sets up a docker network to ease communication between the `mosquitto` broker container and the `mqtt` container.

### How to test?
- Run the containers as per the instructions in the `README.md` file. (**NOTE/NEW**: Don't forget to create the network (instructions in README.md))
- Create a Device.
- View a device. There is a default configuration and a button to edit the device.
- To test the communication with the broker, create a file called `try_broker_publish.py` in the root directory with the following code (to see a successful response, replace `imei_test` with a valid value from one of the devices you created locally):
```
import json
import time
import random

import paho.mqtt.client as mqtt

mqttBroker = "localhost"
client = mqtt.Client("Test")
client.connect(mqttBroker)

imei_test = "135632563278713"


def test_device_logs():
    while True:
        sensor = random.randint(1, 2)
        message = {
            "deviceId": f'SB100{sensor}',
            "dbLevel": random.randint(60, 121),
            "connected": True,
            "longitude": "1.034",
            "latitude": "0.564",
            "batteryLevel": random.randint(10, 100),
            "sigStrength": random.randint(0, 31)
        }
        client.publish("sb/sensor/logs", json.dumps(message))
        print(client.is_connected())
        print("Published message")
        time.sleep(30)


def test_device_config():
    while True:
        message = {
            "imei": imei_test
        }
        client.publish("sb/sensor/configs", json.dumps(message))
        print(f"Published message: {message}")
        time.sleep(10)


def subscribe_to_device_config():
    client.loop_start()
    client.subscribe([(f"sb/sensor/{imei_test}", 1)])

    def print_config(_, __, message):
        print(f"Received message: {message.payload.decode('utf-8')}")

    client.on_message = print_config


subscribe_to_device_config()

test_device_config()
```
- If successful, you should be able to see something similar to this on the command line:
```
Published message: {'imei': '135632563278713'}
Received message: {"configured": 1, "deviceId": "SB1001", "mode": 2, "recLength": 15, "recInterval": 20, "uploadAddr": "http://localhost:8000/audio/"}
```


### Screenshots
![Screenshot 2021-09-01 020952](https://user-images.githubusercontent.com/26762336/131588718-03032b02-9769-4dfb-b076-9f05c1f81518.png)
![Screenshot 2021-09-01 021037](https://user-images.githubusercontent.com/26762336/131588729-454f84d3-cfe1-408a-8d32-cf57d3be39bd.png)
